### PR TITLE
Add compute_template data source.

### PR DESCRIPTION
**Pull request by falzm**
This change introduces a new `compute_template` *data source* allowing
users to dynamically retrieve information about a Compute instance
template, such as its ID or the username.

### CODE REVIEW

File: exoscale/datasource_compute_template.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+	zoneName := d.Get("zone").(string)
+	zone, err := getZoneByName(ctx, client, zoneName)
+	if err != nil {
+		return err
+	}
+
+	resp, err := client.RequestWithContext(ctx, &egoscale.ListTemplates{
+		ZoneID:         zone.ID,
+		TemplateFilter: filter,
+	})
```
Comment (greut) [inline exoscale/datasource_compute_template.go]: :warning: you won't get any pagination here.

**Comment by greut** [comment]
That was much needed, good job.

File: exoscale/datasource_compute_template.go (hunk 1 above)
Comment (falzm) [reply exoscale/datasource_compute_template.go]: Indeed. I don't know what is the page size, but given that we're probably dealing with very few results (I got inspiration from [there](https://github.com/exoscale/terraform-provider-exoscale/blob/master/exoscale/resource_compute.go#L220-L249), is it exposed to the same pagination problem?)

Comment (greut) [reply exoscale/datasource_compute_template.go]: it is but we were in charge of the template list.

**Comment by mcorbin** [comment]
Users should be able to use non-featured templates and to choose the template by ID. I don't think it's possible with the current approach.

**Comment by falzm** [comment]
Original behavior with a non-featured template:

```terraform
resource "exoscale_compute" "test" {
  zone         = "ch-gva-2"
  display_name = "test"
  template     = "Linux CentOS 7.5 64-bit"
  disk_size    = 10
  key_pair     = "default"
}
```

Result:

```console
...
Error: Error applying plan:

1 error(s) occurred:

* exoscale_compute.test: 1 error(s) occurred:

* exoscale_compute.test: Template not found: Linux CentOS 7.5 64-bit (10GB Disk)
```

Using the new data source:

```terraform
locals {
  zone = "ch-gva-2"
}

data "exoscale_compute_template" "mytemplate" {
  zone     = "${local.zone}"
  name     = "Linux CentOS 7.5 64-bit"
  featured = false
}

resource "exoscale_compute" "test" {
  zone         = "${local.zone}"
  display_name = "test"
  template     = "${data.exoscale_compute_template.mytemplate.id}"
  disk_size    = 10
  key_pair     = "default"
}
```

Result:

```console
...
exoscale_compute.test: Creating...
  affinity_group_ids.#: "" => "<computed>"
  affinity_groups.#:    "" => "<computed>"
  disk_size:            "" => "10"
  display_name:         "" => "test"
  gateway:              "" => "<computed>"
  ip4:                  "" => "true"
  ip6:                  "" => "false"
  ip6_address:          "" => "<computed>"
  ip6_cidr:             "" => "<computed>"
  ip_address:           "" => "<computed>"
  key_pair:             "" => "default"
  name:                 "" => "<computed>"
  password:             "<sensitive>" => "<sensitive>"
  security_group_ids.#: "" => "<computed>"
  security_groups.#:    "" => "<computed>"
  size:                 "" => "Medium"
  state:                "" => "<computed>"
  tags.%:               "" => "<computed>"
  template:             "" => "909ce5b7-2132-4ff0-9bf8-aadbb423f7d9"
  user_data_base64:     "" => "<computed>"
  username:             "" => "<computed>"
  zone:                 "" => "ch-gva-2"
exoscale_compute.test: Still creating... (10s elapsed)
exoscale_compute.test: Creation complete after 14s (ID: 1a3cd2a0-0c04-47d2-a266-c5541a973b40)

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
```

Same using a template ID instead of name: 

```terraform
data "exoscale_compute_template" "mytemplate" {
  zone     = "${local.zone}"
  id       = "909ce5b7-2132-4ff0-9bf8-aadbb423f7d9"
  featured = false
}

resource "exoscale_compute" "test" {
  zone         = "${local.zone}"
  display_name = "test"
  template     = "${data.exoscale_compute_template.mytemplate.id}"
  disk_size    = 10
  key_pair     = "default"
}
```

Result:

```console
...
exoscale_compute.test: Creating...
  affinity_group_ids.#: "" => "<computed>"
  affinity_groups.#:    "" => "<computed>"
  disk_size:            "" => "10"
  display_name:         "" => "test"
  gateway:              "" => "<computed>"
  ip4:                  "" => "true"
  ip6:                  "" => "false"
  ip6_address:          "" => "<computed>"
  ip6_cidr:             "" => "<computed>"
  ip_address:           "" => "<computed>"
  key_pair:             "" => "default"
  name:                 "" => "<computed>"
  password:             "<sensitive>" => "<sensitive>"
  security_group_ids.#: "" => "<computed>"
  security_groups.#:    "" => "<computed>"
  size:                 "" => "Medium"
  state:                "" => "<computed>"
  tags.%:               "" => "<computed>"
  template:             "" => "909ce5b7-2132-4ff0-9bf8-aadbb423f7d9"
  user_data_base64:     "" => "<computed>"
  username:             "" => "<computed>"
  zone:                 "" => "ch-gva-2"
exoscale_compute.test: Still creating... (10s elapsed)
exoscale_compute.test: Creation complete after 15s (ID: 466a2cb6-67f8-46f7-85a3-7e1bc003b7b3)

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
```

**Comment by falzm** [comment]
> The only minor downside is that I find the `feature` flag not obvious to get right in the config. How do I know if the template I am looking for is featured or not? Moreover, this could change over time (e.g. when a new template is released) thus break my TF. Would it be possible to remove it and search over all templates?

Well the point was actually to avoid doing "magic" and let the user full control on the template lookup logic. In practice this flag is set to `true` by default because it's the most common use-case, as opposed to the rare times where you want to use a non-featured template (e.g. for a beta version) that are more internal use cases...

**Comment by dajac** [comment]
I see, so what is the official way to know if my template is featured or not?

I think it is no so rare. For instance, I have been using `Linux CentOS 7.5 64-bit` for a very long time and it was featured. One day, it stopped working because Exoscale has release `Linux CentOS 7.6 64-bit` which has become the new featured template. The featured flag has been removed on the previous version. I find this annoying because it took me a very long time before finding out that the feature flag of my template has been removed.

**Comment by mcorbin** [comment]
A solution could be to replace the `featured` parameter by a `filter` parameter (`templatefilter` in Cloudstack). users will be able to choose between `public`, `featured` or `self` templates (and templates provided by Exoscale are always `public` so things will theoretically never break).

Also, having the `filter` parameter will help us when we will refactor the template API (the `featured` filter will probably go away).

**Comment by falzm** [comment]
@mcorbin @dajac I'm OK for replacing the `featured` flag with a `filter` one, however values *public*, *featured* or *self* are confusing and I'm not sure this would be a better alternative...

**Comment by dajac** [comment]
@mcorbin @falzm Using a `filter` might be the best option. It could work well if the options are well explained in the documentation but I would rather use `custom` instead of `self` to stay consistent with the `custom templates`. Also, I would suggest to use `public` be default for the `data source` to avoid breaking "automation code" of customers.

**Comment by falzm** [comment]
@dajac @mcorbin So to sum it up, the idea would be to have a `filter` data source attribute that can accept one of the following value:

* **exoscale**: scope lookup to official Exoscale templates (i.e. performs a CS `listTemplates templatefilter=featured,public` API call)
* **custom**: scope lookup to user-custom templates (i.e performs a CS `listTemplates templatefilter=self` API call)

Did I get it right? If not please provide correction.

**Comment by dajac** [comment]
@falzm Yes, something like this.

**Comment by falzm** [comment]
In a recent custom templates email exchange there has been the following mapping:

> Reminder: flags ⇔ template filter association:
> * Featured == public + featured
> * Community == public + not featured
> * Self == not public

**Comment by pierre-emmanuelJ** [comment]
> @dajac @mcorbin So to sum it up, the idea would be to have a `filter` data source attribute that can accept one of the following value:
> 
> * **exoscale**: scope lookup to official Exoscale templates (i.e. performs a CS `listTemplates templatefilter=featured,public` API call)
> * **custom**: scope lookup to user-custom templates (i.e performs a CS `listTemplates templatefilter=self` API call)
> 
> Did I get it right? If not please provide correction.

I think filtering like that in the [cli](https://github.com/exoscale/cli) too can be a good idea e.g:
automatically register template in self and when you list template show **exoscale**(featured,public) 
and **custom**(self)

**Comment by falzm** [comment]
New version implementing a `filter` attribute, documented as:

- `filter` - Template search filter, must be either `featured` (official
  Exoscale templates), `community` (community-contributed templates) or `mine`
  (custom templates private to my organization). Default is `featured`.

Data source behavior examples:

Using a custom template:

```console
$ exo vm template list --mine
┼──────────────────┼──────┼──────────────┼──────────────────────────────────────┼──────────┼
│ OPERATING SYSTEM │ DISK │ RELEASE DATE │                  ID                  │ CATEGORY │
┼──────────────────┼──────┼──────────────┼──────────────────────────────────────┼──────────┼
│ mybuntu          │ 10   │ 2019-05-03   │ a7e9a35d-11b3-4dd2-b999-75493aca400b │ self     │
┼──────────────────┼──────┼──────────────┼──────────────────────────────────────┼──────────┼
```

```terraform
data "exoscale_compute_template" "mytemplate" {
  zone   = "${local.zone}"
  name   = "mybuntu"
  filter = "mine"
}

output "mytemplate" {
  value = "${data.exoscale_compute_template.mytemplate.id}"
}
```

Return:

```console
...
data.exoscale_compute_template.mytemplate: Refreshing state...

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

mytemplate = a7e9a35d-11b3-4dd2-b999-75493aca400b
```

Using a "community" template:

```console
exo vm template list --community
┼─────────────────────────────────┼──────┼───────────────────┼──────────────────────────────────────┼───────────┼
│        OPERATING SYSTEM         │ DISK │   RELEASE DATE    │                  ID                  │ CATEGORY  │
┼─────────────────────────────────┼──────┼───────────────────┼──────────────────────────────────────┼───────────┼
...
│ Linux CentOS 7.5 64-bit         │      │ 2018-06-21        │ 909ce5b7-2132-4ff0-9bf8-aadbb423f7d9 │ community │
...
┼─────────────────────────────────┼──────┼───────────────────┼──────────────────────────────────────┼───────────┼
```

```terraform
data "exoscale_compute_template" "mytemplate" {
  zone = "${local.zone}"
  name = "mybuntu"
  name     = "Linux CentOS 7.5 64-bit"
  filter = "community"
}

output "mytemplate" {
  value = "${data.exoscale_compute_template.mytemplate.id}"
}

```

Result:

```console
...
data.exoscale_compute_template.mytemplate: Refreshing state...

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

mytemplate = 909ce5b7-2132-4ff0-9bf8-aadbb423f7d9
```

Using a featured template:

```console
exo vm template list --featured
┼───────────────────────────────┼──────┼───────────────────┼──────────────────────────────────────┼──────────┼
│       OPERATING SYSTEM        │ DISK │   RELEASE DATE    │                  ID                  │ CATEGORY │
┼───────────────────────────────┼──────┼───────────────────┼──────────────────────────────────────┼──────────┼
...
│ Linux CentOS 7.6 64-bit       │ 1    │ 2019-03-14        │ 296422c9-c15d-4cca-b599-8d8a48334096 │ 
...
┼───────────────────────────────┼──────┼───────────────────┼──────────────────────────────────────┼──────────┼
```

```terraform
data "exoscale_compute_template" "mytemplate" {
  zone = "${local.zone}"
  name = "mybuntu"
  name     = "Linux CentOS 7.6 64-bit"
  filter = "featured"
}

output "mytemplate" {
  value = "${data.exoscale_compute_template.mytemplate.id}"
}
```

Result:

```console
...
data.exoscale_compute_template.mytemplate: Refreshing state...

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

mytemplate = 296422c9-c15d-4cca-b599-8d8a48334096
```

Referencing a template by its ID also works:

```terraform
data "exoscale_compute_template" "mytemplate" {
  zone = "${local.zone}"
  id     = "296422c9-c15d-4cca-b599-8d8a48334096"
}

output "mytemplate" {
  value = "${data.exoscale_compute_template.mytemplate.name}"
}
```

Result:

```console
...
data.exoscale_compute_template.mytemplate: Refreshing state...

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

mytemplate = Linux CentOS 7.6 64-bit
```

Everybody OK with this? @dajac @mcorbin @brutasse

**Comment by brutasse** [comment]
> Everybody OK with this?

Looks great @falzm :+1:

**Comment by dajac** [comment]
LGMT @falzm!

**Comment by bcachet** [comment]
@pierre-emmanuelJ should be nice to be able to do the same in EXO CLI:
```
exo vm create chtulu --template "MyTemplateName" --filter mine --zone de-fra-1
```

**Comment by falzm** [comment]
> @pierre-emmanuelJ should be nice to be able to do the same in EXO CLI:
> 
> ```
> exo vm create chtulu --template "MyTemplateName" --filter mine --zone de-fra-1
> ```

Sure, feel free to open an issue to suggest this feature in the relevant project ;)

**Pull request merged**